dex/farm/src/lib.rs

```diff
@@ -82,41 +83,69 @@ pub trait Farm:
     #[payable("*")]
     #[endpoint(enterFarm)]
     fn enter_farm_endpoint(&self) -> EnterFarmResultType<Self::Api> {
-        let output_farm_token_payment = self.enter_farm();
         let caller = self.blockchain().get_caller();
+        let output_farm_token_payment = self.enter_farm();
         self.send_payment_non_zero(&caller, &output_farm_token_payment);
         output_farm_token_payment
     }
 
     #[payable("*")]
     #[endpoint(claimRewards)]
-    fn claim_rewards_endpoint(&self) -> ClaimRewardsResultType<Self::Api> {
+    fn claim_rewards_endpoint(&self, opt_orig_caller: OptionalValue<ManagedAddress>) -> ClaimRewardsResultType<Self::Api> {
         let caller = self.blockchain().get_caller();
-        let claim_rewards_result = self.claim_rewards(&caller);
+        let orig_caller = match opt_orig_caller {
+            OptionalValue::Some(opt_caller) => {
+                self.require_sc_address_whitelisted(&caller);
+                opt_caller
+            },
+            OptionalValue::None => caller.clone(),
+        };
+
+        let claim_rewards_result = self.claim_rewards(&orig_caller);
```

Comment:
This will actually call the energy contract further, no?

Reply:
Nope, all the claim rewards logic is inside the farm SC.

Reply:
Actually, yes, this will call `farm-boosted-yields` SC.